File: Linked List/82. Remove Duplicates from Sorted List II/82. Remove Duplicates from Sorted List II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
        # 1. 处理空链表或者只有单个元素的链表
        if not head or not head.next:
            return head

        # 2. 为原始链表加上头部哑结点
        newhead=ListNode(0)
        newhead.next=head
        head=newhead

        # 3. 创建链表游标指针
        pos=head.next     # 初始化链表游标指针为首结点，该游标指针只会指向相同值的结点中的第一个结点
        lastpos=head      # 链表游标指针pos的前驱结点，初始化为哑结点

        while pos and pos.next:    # 若当前游标指针后已经没有结点，或者游标指针本身就是空结点，这时后续不可能再出现相同值的结点，因此直接退出处理循环
            # 3.1 若当前结点的值和后驱结点的值不相等，则无需进行删除，直接将pos指向下一个新值的结点
            if pos.val!=pos.next.val: 
                lastpos=pos       # 更新指针
                pos=pos.next
            
            # 3.2 若当前结点的值和后驱结点的值相等，则需要进行删除，这时先找到第一个和当前结点值不相等的结点
            else:
                now=pos.next     
                while now and now.val==pos.val:   # 注意第一个条件，后续可能无法找到与pos不相等的结点
                    now=now.next

                lastpos.next=now      # 将当前结点pos的前驱结点lastpos和第一个和当前结点值不相等的结点now相连接

                # python 无需进行手动的内存释放，因此可以省略释放这些重复结点的内存的过程

                lastpos=lastpos       # 更新指针
                pos=now 
        
        return head.next

                

File: Linked List/82. Remove Duplicates from Sorted List II/82. Remove Duplicates from Sorted List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

//本题中要求删除掉所有出现重复的值对应的结点，因此需要特别注意一个边缘问题——有可能整个链表删除后变成空链表

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        // 0. 处理特殊情况
        if(!head||!head->next)
        return head;

        // 1. 因为可能需要删除原链表的头结点，因此直接先考虑给原链表头部加上哑结点
        ListNode *newhead=new ListNode(0);
        newhead->next=head;
        head=newhead;

        // 1. 创建链表游标指针
        ListNode *pos=head->next;  //链表游标指针pos
        ListNode *lastpos=head;    //记录链表游标指针的pos上一个取值
        
        // 2. 使用链表游标指针遍历整个链表，直到到达链表的最后一个结点
        while(pos!=NULL&&pos->next!=NULL)     //游标指针pos到最后一个结点就可以停止迭代过程
        {
            // 2.1 不存在相同值的结点，则直接更新指针进行下一轮迭代
            if(pos->next->val!=pos->val)   
            {
                lastpos=pos;
                pos=pos->next;
                continue;
            }
            // 2.2  存在相同值的结点，则进行删除处理
            else
            {
                // 2.2.1 先查找pos后的第一个不等于pos的值的结点now
                ListNode *now=pos->next;
                while(now&&now->val==pos->val)
                {
                    now=now->next;
                }

                //2.2.2 释放待删除结点的内存
                ListNode *temp=lastpos->next;
                while(temp!=now)
                {
                    ListNode *temp2=temp;
                    temp=temp->next;
                    delete temp2;
                }

                // 2.2.3 将lastpos和结点now相连
                lastpos->next=now;

                // 2.2.4 更新指针pos和lastpos
                lastpos=lastpos;    //lastpos不变
                pos=now;
            }
        }

        ListNode *temp=head;
        head=head->next;
        delete temp;

        return head;
    }
};